metadeploy/api/views.py

```diff
@@ -176,51 +191,145 @@ def filter_get_one(self, qs):
     def preflight_get(self, request):
         plan = get_object_or_404(Plan.objects, id=self.kwargs["pk"])
         preflight = PreflightResult.objects.most_recent(
-            user=request.user, plan=plan, is_valid_and_complete=False
+            scratch_org_id=request.session.get("scratch_org_id", None),
+            user=request.user,
+            plan=plan,
+            is_valid_and_complete=False,
         )
         if preflight is None:
             return Response("", status=status.HTTP_404_NOT_FOUND)
         serializer = PreflightResultSerializer(instance=preflight)
         return Response(serializer.data)
 
     def preflight_post(self, request):
+        scratch_org_id = request.session.get("scratch_org_id", None)
         plan = get_object_or_404(Plan.objects, id=self.kwargs["pk"])
         is_visible_to = plan.is_visible_to(
             request.user
         ) and plan.version.product.is_visible_to(request.user)
-        if not is_visible_to:
+        user_kwargs = {
+            "user": request.user if request.user.is_authenticated else None,
+            "uuid": scratch_org_id,
+        }
+        if not is_visible_to or not any(user_kwargs.values()):
             return Response("", status=status.HTTP_403_FORBIDDEN)
+
+        forced_user_kwargs = {}
+        try:
+            scratch_org_job = ScratchOrgJob.objects.get(uuid=scratch_org_id)
+            config = scratch_org_job.config
+            org_kwargs = {
+                "organization_url": config["instance_url"],
+                "org_id": config["org_id"],
+                # @davisagli suspected we'd need this, but it's not
+                # currently used anywhere in jobs.py, so I am not sure
+                # why we would:
+                # "username": config["username"],
```

Comment:
@wlonk you're right. I was thinking we needed it to get an access token, but instead run_flows is constructing an OrgConfig using the access_token that was obtained via oauth when the user logged in. That's obviously not going to work for scratch orgs because we don't have a logged in user or token from oauth. But we do have an access token in the ScratchOrgJob config that was obtained in _get_access_token after creation of the scratch org.

We should also make sure that this access token is removed from the ScratchOrgJob on a similar schedule to when we delete the oauth tokens (once the install job is complete, or if the ScratchOrgJob has existed for N minutes without running an install job on it.

Reply:
I split this out into a new card: https://trello.com/c/vjQsgenR